app/components/form_components/autocomplete/sort.js

```diff
@@ -0,0 +1,48 @@
+// Returns an array of integers representing the position of each regex match.
+const matchPositions = (string, regexes) => regexes.map(regex => string.search(regex)).filter(i => i >= 0)
+
+const calculateWeight = (name, rawSynonyms, query) => {
+  const regexes = query.split(/\s+/).map(word => new RegExp('\\b' + word, 'i'))
+  const synonyms = rawSynonyms || []
+
+  const nameMatchPositions = matchPositions(name, regexes)
+  const synonymMatchPositions = synonyms.map(synonym => matchPositions(synonym, regexes)).flat()
+
+  // Require all parts of the query to be matched:
+  if (nameMatchPositions.length !== regexes.length && synonymMatchPositions.length !== regexes.length) return 0
+
+  // Case insensitive exact matches:
+  const nameIsExactMatch = query === name.toLowerCase()
+  const synonymIsExactMatch = synonyms.some(s => query === s.toLowerCase())
+
+  // Case insensitive 'starts with':
+  const nameStartsWithQuery = nameMatchPositions.includes(0)
+  const synonymStartsWithQuery = synonymMatchPositions.includes(0)
+  const wordInNameStartsWithQuery = nameMatchPositions.length > 0
```

Comment:
Should there be a matching `wordInSynonymStartsWithQuery`?

Reply:
@edwardhorsford - for the sake of moving things along, are you ok with me leaving this rule out for now until we have multi-word synonyms to test with in the service?

I can add a ticket to the ideas board along with the work for adding synonyms like we discussed (maths, masters)